### PR TITLE
Feat/number input 2

### PR DESCRIPTION
**Pull request by M-i-k-e-l**
## Description
NumberInput - a new take on the component
Update the UI in `onBlur`
Removed tests from the npm package

WOAUILIB-2640

## Changelog
NumberInput - new component (⚠️ also removed tests from the npm package)

### CODE REVIEW

**Review by lidord-wix — COMMENTED** [review]
@M-i-k-e-l 
Just before I'll review the code: 
1. Looks very good!
2. When entering a large number the `NumberInput` stretches its width to the screen width (and more) and in the guidelines they ask to keep an s5 margin from each side.
3. Just to make it easier to review - how different is this PR's implementation from the previous one? should I review all the files? is there something I should focus on more?

**Comment by M-i-k-e-l** [comment]
> @M-i-k-e-l Just before I'll review the code:

Please ping me in another way if you think you are blocked from reviewing because when I get an email with a comment I assume you've already reviewed it and I may not immediately look at the comments themselves, this delays the release on the component...

> 1. Looks very good!
> 2. When entering a large number the `NumberInput` stretches its width to the screen width (and more) and in the guidelines they ask to keep an s5 margin from each side.
> 3. Just to make it easier to review - how different is this PR's implementation from the previous one? should I review all the files? is there something I should focus on more?

1. 10x!
2. Fixed (partly in the component and partly in the screen); there's still a bug with a long text where the `trailingAccessory` is pushed, this is a bug in `TextField` [see here](https://github.com/wix/react-native-ui-lib/pull/2345#discussion_r1035944834).
3. Very different.

File: src/components/numberInput/Presenter.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import {isEmpty} from 'lodash';
+
+export type NumberInputResult =
```
Comment (lidord-wix) [inline src/components/numberInput/Presenter.ts:3]: Maybe `NumberInputData` or `NumberInputValue`?

File: src/components/numberInput/index.tsx [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+  const _containerStyle = useMemo(() => {
+    return [styles.containerStyle, containerStyle];
+  }, [containerStyle]);
+
+  const _onChangeText = useCallback((text: string) => {
+    const newData = parseInput(text, options);
+    onChangeNumber(newData);
+    setData(newData);
+    onChangeText?.(text);
```
Comment (lidord-wix) [inline src/components/numberInput/index.tsx:133]: Do we need both `onChangeNumber` and `onChangeText`? 
maybe we can omit onChangeText from the type..

**Comment by M-i-k-e-l** [comment]
> I wrote some comments on the code (it looks like a lot but many of them are about naming and small stuff), and more general comments:
> 
> 1. I still see an issue with a long input - you can't see the last char you typed (it's cut), you can also see that the left margin is kept when the right one is not.

Yes, I wrote about this in the previous section (number 3 [here](https://github.com/wix/react-native-ui-lib/pull/2333#issuecomment-1337459492))

> 2. If you do keep the two example screens, consider adding another file for all the shared code (processInput, onChangeNumber, validation functions, and so on).

I think we want to change the screens, but I am not sure about shared code in screens, it can make the example less clear which might be more important.

> 3. If doing changes as I suggested, don't forget to update the relevant API files.

ok

> 4. BTW, I already wrote some of the comments I added here in the previous PR.. that's why I asked about the diffs before I did the review :)

ok

File: src/components/numberInput/index.tsx [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+export {NumberInputResult};
+
+export type NumberInputProps = React.PropsWithRef<
+  Omit<TextFieldProps, 'leadingAccessory' | 'trailingAccessory' | 'value'> & ThemeComponent
+> & {
+  /**
+   * Callback that is called when the number value has changed (undefined in both if the user has deleted the number).
+   */
+  onChangeNumber: (result: NumberInputResult) => void;
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/numberInput/index.tsx:17]: I don't like `value` because I think that's more like a single data (`number` \ `string` etc), I can't say I like `data` but I'll trust you on this.

File: demo/src/screens/componentScreens/NumberInputScreen/NumberInputErrorOnChangeScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  View,
+  Typography,
+  Constants,
+  Incubator
+} from 'react-native-ui-lib';
+
+const NumberInputErrorOnChangeScreen = () => {
+  const currentData = useRef<NumberInputResult>();
+  const minimum = useRef<number>(5000);
+  // const [initialNumber, setInitialNumber] = useState<number>(100);
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply demo/src/screens/componentScreens/NumberInputScreen/NumberInputErrorOnChangeScreen.tsx:17]: Yes, I had another example which is now gone, removed

File: src/components/numberInput/index.tsx [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+  /**
+   * A valid number (in en locale, i.e. only digits and a decimal point).
+   */
+  initialNumber?: number;
+  /**
+   * Number of digits after the decimal point. Must be in the range 0 - 20, inclusive.
+   */
+  fractionDigits?: number;
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/numberInput/index.tsx:25]: Done, the `locale` is not working properly but fixing it will probably require more data to be saved and I'm not sure it's worth it

File: src/components/numberInput/index.tsx [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  }, []);
+
+  useDidUpdate(() => {
+    handleInitialValueChange();
+  }, [initialNumber]);
+
+  const hasText = useMemo(() => {
+    // Render both leading and trailing accessories so the text is centered AND the margin between the text and the accessories is correct
+    return !isEmpty(leadingText) || !isEmpty(trailingText);
+  }, [leadingText, trailingText]);
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/numberInput/index.tsx:99]: That's a big `if` :)
We can talk about it

File: src/components/numberInput/index.tsx [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    onChangeText?.(text);
+  },
+  [onChangeNumber, options, onChangeText]);
+
+  const value = useMemo(() => {
+    return data?.type === 'valid' || data?.type === 'error' ? data.userInput : '';
+  }, [data]);
+
+  const formatter = useCallback(() => {
+    return data?.type === 'valid' ? data.formattedNumber : data?.type === 'error' ? data.userInput : '';
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/numberInput/index.tsx:142]: That results with a TS error (because `data` is a Discriminated Union, which seems to not fully support this (yet?)).

File: src/components/numberInput/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (lidord-wix) [reply src/components/numberInput/index.tsx:17]: Yes

File: src/components/numberInput/index.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (lidord-wix) [reply src/components/numberInput/index.tsx:110]: Let's :)

File: src/components/numberInput/index.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (lidord-wix) [reply src/components/numberInput/index.tsx:150]: So you can add `userInput` to the empty state as well, WDYT?

**Comment by lidord-wix** [comment]
> > I wrote some comments on the code (it looks like a lot but many of them are about naming and small stuff), and more general comments:
> > 
> > 1. I still see an issue with a long input - you can't see the last char you typed (it's cut), you can also see that the left margin is kept when the right one is not.
> 
> Yes, I wrote about this in the previous section (number 2 [here](https://github.com/wix/react-native-ui-lib/pull/2333#issuecomment-1337459492))
> 
Oh I thought you only wrote about the case with a `trailingAccessory`
> > 2. If you do keep the two example screens, consider adding another file for all the shared code (processInput, onChangeNumber, validation functions, and so on).
> 
> I think we want to change the screens, but I am not sure about shared code in screens, it can make the example less clear which might be more important.
> 
> > 3. If doing changes as I suggested, don't forget to update the relevant API files.
> 
> ok
> 
you forgot :) 
> > 4. BTW, I already wrote some of the comments I added here in the previous PR.. that's why I asked about the diffs before I did the review :)
> 
> ok

**Comment by lidord-wix** [comment]
@M-i-k-e-l 
I think you missed some of my comments, please take a look again :)

File: demo/src/screens/ExampleScreenPresenter.tsx [hunk 8]
```diff
@@ -35,6 +35,46 @@ export function renderHeader(title: string, others?: TextProps) {
...
+      )}
+      <SegmentedControl
+        initialIndex={index}
+        segments={options}
+        onChangeIndex={index => setState(options[index].value)}
+      />
+    </View>
+  );
+}
+
```
Comment (lidord-wix) [inline demo/src/screens/ExampleScreenPresenter.tsx:77]: This is already a huge PR and this change shouldn't be part of it.
next time please do it in a separate PR 🙏🏽

File: src/incubator/TextField/index.tsx [hunk 9]
```diff
@@ -99,15 +99,19 @@ const TextField = (props: InternalTextFieldProps) => {
   const hidePlaceholder = shouldHidePlaceholder(props, fieldState.isFocused);
   const retainTopMessageSpace = !floatingPlaceholder && isEmpty(trim(label));
   const centeredContainerStyle = centered && styles.centeredContainer;
-  const centeredLabelStyle = centered && styles.centeredLabel;
+  const _labelStyle = useMemo(() => (centered ? [labelStyle, styles.centeredLabel] : labelStyle),
+    [labelStyle, centered]);
+  const _validationMessageStyle = useMemo(() => {
+    return centered ? [validationMessageStyle, styles.centeredValidationMessage] : validationMessageStyle;
+  }, [validationMessageStyle, centered]);
```
Comment (lidord-wix) [inline src/incubator/TextField/index.tsx:106]: Also here, 
This change shouldn't be part of this PR.. 
please do it in a different PR next time 🙏🏽

File: src/components/numberInput/index.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/numberInput/index.tsx:150]: I don't really like it, it'll force empty to be inserted as:
```
export const EMPTY: NumberInputData = {type: 'empty', userInput: ''};
```
Instead of:
```
export const EMPTY: NumberInputData = {type: 'empty'};
```

BTW, here we still need the `data.formattedNumber` AFAIK

File: demo/src/screens/ExampleScreenPresenter.tsx [hunk 10]
```diff
@@ -35,6 +35,46 @@ export function renderHeader(title: string, others?: TextProps) {
   );
 }
 
+// eslint-disable-next-line max-params
+export function renderBooleanOptionForFunction(title: string,
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply demo/src/screens/ExampleScreenPresenter.tsx]: Done

File: src/components/numberInput/index.tsx [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+  trailingTextStyle?: StyleProp<ViewStyle>;
+  /**
+   * The margin to the left of the input
+   */
+  marginLeft?: number;
+  /**
+   * The margin to the right of the input
+   */
+  marginRight?: number;
+};
```
[the remark this answers is not in the document]
Comment (lidord-wix) [reply src/components/numberInput/index.tsx]: Please update the api file as well :)

**Review by lidord-wix — APPROVED** [review]
Approved, see my last two comments :)
